File: Day19-BinaryTree6/450. Delete Node in a BST/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    
    def printTree(self, root):
        if root:
            print(root.val)
            self.printTree(root.left)
            self.printTree(root.right)



class Solution:
    def deleteNode(self, root: TreeNode, key: int)->TreeNode:
        if root is None:
            return root
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            if root.left == None:
                return root.right
            elif root.right == None:
                return root.left
            else:
                root.val = self.successor(root)
                root.right = self.deleteNode(root.right, root.val)
        return root

    def successor(self, root):
        root = root.right
        while root.left:
            root = root.left
        return root.val
    
def main():
    root = TreeNode(5)
    root.left = TreeNode(3)
    root.right = TreeNode(6)
    root.left.left = TreeNode(2)
    root.left.right = TreeNode(4)
    root.right.right = TreeNode(7)
    obj = Solution()
    obj.deleteNode(root, 3)
    root.printTree(root)

if __name__ == "__main__":
    main()

File: Day19-BinaryTree6/450. Delete Node in a BST/main.cpp
#include <iostream>

using namespace std;

struct TreeNode{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(): val(0), left(nullptr), right(nullptr){}
    TreeNode(int x): val(x), left(nullptr), right(nullptr){}
    TreeNode(int x, TreeNode* left, TreeNode* right): val(x), left(left), right(right){}
};

void preOrder(TreeNode* root){
    if(root == nullptr) return;
    cout << root->val << " ";
    preOrder(root->left);
    preOrder(root->right);
}

class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(root == nullptr) return nullptr;
        if(key < root->val){
            root->left = deleteNode(root->left, key);
        }else if(key > root->val){
            root->right = deleteNode(root->right, key);
        }else{
            if(root->left == nullptr){
                return root->right;
            }else if(root->right == nullptr){
                return root->left;
            }else{
                TreeNode* minNode = getMin(root->right);
                root->val = minNode->val;
                root->right = deleteNode(root->right, minNode->val);
            }
        }
        return root;
    }

    TreeNode* getMin(TreeNode* node){
        while(node->left != nullptr){
            node = node->left;
        }
        return node;
    }
    TreeNode* successor(TreeNode* root){
        root = root->right;
        while(root->left != nullptr){
            root = root->left;
        }
        return root;
    }
    
};

int main(){
    TreeNode* root = new TreeNode(5);
    root->left = new TreeNode(3);
    root->right = new TreeNode(6);
    root->left->left = new TreeNode(2);
    root->left->right = new TreeNode(4);
    root->right->right = new TreeNode(7);
    Solution s;
    TreeNode* res = s.deleteNode(root, 3);
    preOrder(res);
    return 0;
}